packages/stellar-axelar-std-derive/src/storage.rs

```diff
@@ -0,0 +1,400 @@
+use heck::ToSnakeCase;
+use itertools::Itertools;
+use proc_macro2::{Ident, TokenStream};
+use quote::{format_ident, quote};
+use syn::{Attribute, Data, DataEnum, DeriveInput, Fields, FieldsNamed, Meta, Type, Variant};
+
+struct StorageAttributes {
+    storage_type: StorageType,
+    value_type: Type,
+}
+
+#[derive(Debug)]
+enum StorageType {
+    Instance,
+    Persistent,
+    Temporary,
+}
+
+impl StorageType {
+    pub fn storage_method(&self) -> TokenStream {
+        match self {
+            Self::Instance => quote! { instance },
+            Self::Persistent => quote! { persistent },
+            Self::Temporary => quote! { temporary },
+        }
+    }
+
+    pub fn ttl_method(&self) -> TokenStream {
+        match self {
+            Self::Persistent => quote! {
+                stellar_axelar_std::ttl::extend_persistent_ttl(env, &key);
+            },
+            Self::Instance => quote! {
+                stellar_axelar_std::ttl::extend_instance_ttl(env);
+            },
+            Self::Temporary => quote! {},
+        }
+    }
+}
+
+/// Generates the storage enum and its associated functions.
+pub fn contractstorage(input: &DeriveInput) -> TokenStream {
+    let name = &input.ident;
+
+    let Data::Enum(DataEnum { variants, .. }) = &input.data else {
+        panic!("contractstorage can only be used on enums.");
+    };
+
+    let transformed_variants: Vec<_> = variants.iter().map(transform_variant).collect();
+
+    let storage_fns: Vec<_> = variants
+        .iter()
+        .map(|variant| {
+            let storage_type = storage_type(&variant.attrs);
+            let value_type = value_type(&variant.attrs);
+            storage_fns(
+                name,
+                variant,
+                &StorageAttributes {
+                    storage_type,
+                    value_type,
+                },
+            )
+        })
+        .collect();
+
+    let public_fns: Vec<_> = variants
+        .iter()
+        .map(|variant| {
+            let value_type = value_type(&variant.attrs);
+            public_storage_fns(name, variant, &value_type)
+        })
+        .collect();
+
+    let output = quote! {
+        #[contracttype]
+        enum #name {
+            #(#transformed_variants,)*
+        }
+
+        impl #name {
+            #(#storage_fns)*
+        }
+
+        #(#public_fns)*
+    };
```

Comment:
I'd like to see an auto-generated test block for every storage type, so that any changes to the actual DataKey enum are flagged, since a migration would likely be needed
```rust
#[cfg(test)]
mod test {
    #[test]
    fn ensure_storage_schema_is_unchanged() {
    }
}
```